controller/src/main/java/io/pravega/controller/store/stream/ZkOrderedStore.java

```diff
@@ -92,30 +95,54 @@
                                 entity.getBytes(Charsets.UTF_8))
                                    .thenCompose(positionPath -> {
                                        int position = getPositionFromPath(positionPath);
                                        if (position > rollOverAfter) {
-                                           // if newly created position exceeds rollover limit, we need to delete that entry 
-                                           // and roll over. 
-                                           // 1. delete newly created path
-                                           return storeHelper.deletePath(positionPath, false)
-                                                             // 2. seal latest collection
-                                                             .thenCompose(v -> storeHelper.createZNodeIfNotExist(
-                                                                     getCollectionSealedPath(scope, stream, latestcollectionNum)))
-                                                             // 3. call addEntity recursively
-                                                             .thenCompose(v -> addEntity(scope, stream, entity))
-                                                             // 4. delete empty sealed collection path
-                                                             .thenCompose(orderedPosition -> 
-                                                                     tryDeleteSealedCollection(scope, stream, latestcollectionNum)
-                                                                     .thenApply(v -> orderedPosition));
+                                           // if newly created position exceeds rollover limit, we need to delete that entry
+                                           // and roll over.
+                                           try {
+                                               // prepare transaction: delete newly created path, create new collection and put the entity in
+                                               int newCollectionNumber = latestCollectionNumber + 1;
+
+                                               String collectionPath = getCollectionPath(scope, stream, newCollectionNumber);
+                                               String entitiesPath = getEntitiesPath(scope, stream, newCollectionNumber);
+                                               String entitySequentialPath = getEntitySequentialPath(scope, stream, newCollectionNumber);
+
+                                               CuratorOp op1 = storeHelper.getClient().transactionOp().delete().forPath(positionPath);
+                                               CuratorOp op2 = storeHelper.getClient().transactionOp().create().forPath(collectionPath);
+                                               CuratorOp op3 = storeHelper.getClient().transactionOp().create().forPath(entitiesPath);
```

Comment:
same question as above. 

Note: there could be multiple concurrent attempts to add new entities to the collection. 
So if a new collection had to be created, then the parent paths here could clash and we could get DataExists for these two parent paths that are being attempted here.